File: frontend/src/features/local-models/schemas.ts
import { z } from "zod";

export const LocalModelStatusSchema = z.object({
  bridgeRunning: z.boolean(),
  ollamaRunning: z.boolean(),
  endpoint: z.string().nullable().optional().default(null),
  loadedCount: z.number().int().nonnegative(),
});

export const LocalModelSchema = z.object({
  name: z.string(),
  digest: z.string().nullable().optional().default(null),
  sizeBytes: z.number().int().nonnegative(),
  modifiedAt: z.string().datetime({ offset: true }).nullable().optional().default(null),
  loaded: z.boolean(),
  loadedSizeBytes: z.number().int().nonnegative(),
  lastUsedAt: z.string().datetime({ offset: true }).nullable().optional().default(null),
});

export const LocalModelListSchema = z.object({
  models: z.array(LocalModelSchema),
});

export const LocalModelMetricsSchema = z.object({
  localTps: z.number(),
  requestCount: z.number().int().nonnegative(),
  queueDepth: z.number().int().nonnegative(),
  quotaSavedTokens: z.number().int().nonnegative(),
  quotaSavedPercent: z.number(),
});

export type LocalModelStatus = z.infer<typeof LocalModelStatusSchema>;
export type LocalModel = z.infer<typeof LocalModelSchema>;
export type LocalModelList = z.infer<typeof LocalModelListSchema>;
export type LocalModelMetrics = z.infer<typeof LocalModelMetricsSchema>;

File: app/modules/local_models/schemas.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class LocalModelStatusResponse(DashboardModel):
    bridge_running: bool = False
    ollama_running: bool = False
    endpoint: str | None = None
    loaded_count: int = 0


class LocalModelEntry(DashboardModel):
    name: str
    digest: str | None = None
    size_bytes: int = 0
    modified_at: datetime | None = None
    loaded: bool = False
    loaded_size_bytes: int = 0
    last_used_at: datetime | None = None


class LocalModelListResponse(DashboardModel):
    models: list[LocalModelEntry] = Field(default_factory=list)


class LocalModelMetricsResponse(DashboardModel):
    local_tps: float = 0.0
    request_count: int = 0
    queue_depth: int = 0
    quota_saved_tokens: int = 0
    quota_saved_percent: float = 0.0
